### ISSUE TITLE
Suggestion: Adding Learning Curve Plot Function to the package

### ISSUE THREAD

**Issue by hxd1011** [topepo/caret#278]
Hi, There, 

This is Haitao. I am a data scientist and I really love the caret package.

I have one suggestion of making it better: adding the "learning curve plot" to the package. When I say learning curve I am referring to Andrew NG's term

https://www.youtube.com/watch?v=IRyZ-UT04qg

This is a great tool for diagnosis, I personally think this is even more important than parameter tuning. In most cases, even using 10% of the data will have the similar performance of using all of the data (especially for high bias models, such as linear)

I am thinking that we are wasting a lot of time to tune the parameter using all data and complicated cross validation. Adding learning curve plot will save us a lot of time from beginning.

Feel free to contact me for any questions. 

Thanks.

**Comment by topepo** [comment]
I never really liked that visualization since it does a ton of computing to tell you how much computing you should have done. However, I'm really procrastinating today. 

A draft function and some code is below. It uses the same test set across all the model fits and the resampled values reflect the optimal tuning parameter values (if the model tuned across multiple values). 

Please test it out and see where it breaks. It is difficult to write a function like this and have it work across almost every possible use case so the more testing, the better.

```r
##        dat: entire data set used for modling
##          y: character stirng for the outcome column name
## proportion: proportion of data used to train the model
##  test_prop: proportion of data used initially set aside for testing
##    verbose: write out a log of training milestones
##        ...: arguments to pass to `train`

learing_curve_dat <- function(dat, 
                              outcome = colnames(dat)[1],
                              proportion = (1:10)/10, test_prop = 0, 
                              verbose = TRUE, ...) {
  
  proportion <- sort(unique(proportion))
  n_size <- length(proportion)
  
  if(test_prop > 0) {
    for_model <- createDataPartition(dat[, outcome], p = 1 - test_prop, list = FALSE)
  } else for_model <- 1:nrow(dat)
  
  n <- length(for_model)
  
  resampled <- vector(mode = "list", length = n_size)
  tested <- if(test_prop > 0) resampled else NULL
  apparent <- resampled
  for(i in seq(along = proportion)) {
    if(verbose) cat("Training for ", round(proportion[i]*100, 1), 
                    "% (n = ", floor(n*proportion[i]), ")\n", sep = "")
    in_mod <- if(proportion[i] < 1) sample(for_model, size = floor(n*proportion[i])) else for_model
    mod <- train(x = dat[in_mod, colnames(dat) != outcome, drop = FALSE],
                 y = dat[in_mod, outcome],
                 ...)
    if(i == 1) perf_names <- mod$perfNames
    resampled[[i]] <- merge(mod$resample, mod$bestTune)
    resampled[[i]]$Training_Size <- length(in_mod)
    
    if(test_prop > 0) {
      if(!mod$control$classProbs) {
        test_preds <- extractPrediction(list(model = mod), 
                                        testX = dat[-for_model, colnames(dat) != outcome, drop = FALSE],
                                        testY = dat[-for_model, outcome])
      } else {
        test_preds <- extractProb(list(model = mod), 
                                  testX = dat[-for_model, colnames(dat) != outcome, drop = FALSE],
                                  testY = dat[-for_model, outcome])
      }
      test_perf <- mod$control$summaryFunction(test_preds, lev = mod$finalModel$obsLevels)
      test_perf <- as.data.frame(t(test_perf))
      test_perf$Training_Size <- length(in_mod)
      tested[[i]] <- test_perf
      try(rm(test_preds, test_perf), silent = TRUE)
    }
    
    if(!mod$control$classProbs) {
      app_preds <- extractPrediction(list(model = mod), 
                                     testX = dat[in_mod, colnames(dat) != outcome, drop = FALSE],
                                     testY = dat[in_mod, outcome])
    } else {
      app_preds <- extractProb(list(model = mod), 
                               testX = dat[in_mod, colnames(dat) != outcome, drop = FALSE],
                               testY = dat[in_mod, outcome])
    }
    app_perf <- mod$control$summaryFunction(app_preds, lev = mod$finalModel$obsLevels)
    app_perf <- as.data.frame(t(app_perf))
    app_perf$Training_Size <- length(in_mod)    
    apparent[[i]] <- app_perf
    
    try(rm(mod, in_mod, app_preds, app_perf), silent = TRUE)
  }
  
  resampled <- do.call("rbind", resampled)
  resampled <- resampled[, c(perf_names, "Training_Size")]
  resampled$Data <- "Resampling"
  apparent <- do.call("rbind", apparent)
  apparent <- apparent[, c(perf_names, "Training_Size")]
  apparent$Data <- "Training"
  out <- rbind(resampled, apparent)
  if(test_prop > 0) {
    tested <- do.call("rbind", tested)
    tested <- tested[, c(perf_names, "Training_Size")]
    tested$Data <- "Testing"
    out <- rbind(out, tested)
  }
  out
}

###################################################################
## Classification example

library(caret)

set.seed(1412)
class_dat <- twoClassSim(1000)

set.seed(29510)
lda_data <- learing_curve_dat(dat = class_dat, outcome = "Class",
                              test_prop = 1/4, 
                              ## `train` arguments
                              method = "lda", 
                              metric = "ROC",
                              trControl = trainControl(classProbs = TRUE, 
                                                       summaryFunction = twoClassSummary))

ggplot(lda_data, aes(x = Training_Size, y = ROC, color = Data)) + 
  geom_smooth(method = loess, span = .8) + 
  theme_bw()


set.seed(29510)
rf_data <- learing_curve_dat(dat = class_dat, outcome = "Class",
                             test_prop = 1/4, 
                             ## `train` arguments
                             method = "rf", 
                             metric = "ROC",
                             tuneLength = 4,
                             trControl = trainControl(classProbs = TRUE, 
                                                      method = "cv",
                                                      summaryFunction = twoClassSummary))

ggplot(rf_data, aes(x = Training_Size, y = ROC, color = Data)) + 
  geom_smooth(method = loess, span = .8) + 
  theme_bw()


###################################################################
## Regression example


set.seed(19135)
reg_dat <- SLC14_1(1000)

set.seed(31535)
bag_data <- learing_curve_dat(dat = reg_dat, outcome = "y",
                              test_prop = 1/4, 
                              ## `train` arguments
                              method = "treebag", 
                              trControl = trainControl(method = "cv"),
                              ## `bagging` arguments
                              nbagg = 100)

ggplot(bag_data, aes(x = Training_Size, y = RMSE, color = Data)) + 
  geom_smooth(method = loess, span = .8) + 
  theme_bw()


set.seed(31535)
svm_data <- learing_curve_dat(dat = reg_dat, outcome = "y",
                              test_prop = 1/4, 
                              ## `train` arguments
                              method = "svmRadial", 
                              preProc = c("center", "scale"),
                              tuneGrid = data.frame(sigma =  0.03, C = 2^10),
                              trControl = trainControl(method = "cv"))

ggplot(svm_data, aes(x = Training_Size, y = RMSE, color = Data)) + 
  geom_smooth(method = loess, span = .8) + 
  theme_bw()



set.seed(31535)
svm_no_test <- learing_curve_dat(dat = reg_dat, outcome = "y",
                                 test_prop = 0, 
                                 ## `train` arguments
                                 method = "svmRadial", 
                                 preProc = c("center", "scale"),
                                 tuneGrid = data.frame(sigma =  0.03, C = 2^10),
                                 trControl = trainControl(method = "cv"))

ggplot(svm_no_test, aes(x = Training_Size, y = RMSE, color = Data)) + 
  geom_smooth(method = loess, span = .8) + 
  theme_bw()
```

**Comment by hxd1011** [comment]
Wow, nice work topepo, thank you very much for the feedback. It runs well on my machine and I am planning to test it on other models.

What do you think about next step? Should we work together to add it into the library?

Thanks again

**Comment by topepo** [comment]
I'll put it in for the next version but I probably won't get to it until just before the next release. 

Please test as much as you can before then and provide comments/changes that you think are needed.

**Comment by hxd1011** [comment]
Thanks, I will carefully test it, if you want I can also write a Vignettes to illustrate how to use it to save time on model exploring.

**Comment by jjalcolea** [comment]
Hi! Good to see that this is already on its way into caret. Any idea on when will it "officially" be available?

Thanks and best wishes,

Juanje.

**Comment by topepo** [comment]
It's already there. See `?learing_curve_dat`.

**Comment by sajidzaidi** [comment]
i believe this is misspelled? should be learning_curve_dat

**Comment by lbcommer** [comment]
It looks like it doesn't work with linear regression models with formula.
Is there any way to have learning curves with those models?
This produces an error:

`library(mlbench)
data(BostonHousing)
fit <- learing_curve_dat(dat = BostonHousing, outcome = "mdev",
                              ## `train` arguments
                              form = formula("medv ~ ."),
                              method = "lm", 
                              metric = "RMSE"  
                          )`

**Comment by juliangkr** [comment]
I might misunderstand the code, but for me it seems like that there is a bug.
These lines:

``` R
if(!mod$control$classProbs) {
        test_preds <- extractPrediction(list(model = mod), 
                                        testX = dat[-for_model, colnames(dat) != outcome, drop = FALSE],
                                        testY = dat[-for_model, outcome])
      } else {
        test_preds <- extractProb(list(model = mod), 
                                  testX = dat[-for_model, colnames(dat) != outcome, drop = FALSE],
                                  testY = dat[-for_model, outcome])
      }
      test_perf <- mod$control$summaryFunction(test_preds, lev = mod$finalModel$obsLevels)
```

should return the predictions of the model on the test set, correct? But as far as I can see, the function returns the predictions for both, the training and the test set. 
Afterwards, in the next lines the performance is calculated on all the predictions. However, these are wrong, as all predictions are used, both the predictions on the test and training data? 
If I am wrong, could someone please explain me what I am doing wrong, because I can't reach the performances from the learning models using predict.train()

**Comment by topepo** [comment]
> the function returns the predictions for both, the training and the test set.

That is what is described in the `Details` section of the help page. 

@lbcommer and @juliangkr: it would be a good idea to start new issues instead of added to one that is already closed. It is very easy for appended issues to slip through the cracks. Please read the posting guide when you make new issues.

**Comment by top-engineer** [comment]
The fact that what is "claimed" to be a solid machine learning toolbox ("caret") does yet to have a function that plots learning curve is so ridiculous. The learnings_curve_dat() function is poorly documented and seems disconnected from the train() object and falls too short to meet the basic requirement of visualizing bias, variance and irreducible error for different training set sizes.

I hope the caret authors would consider giving these kind of diagnostics functions some serious ravamp.
